### PR TITLE
replace `request` with `node-fetch`

### LINKED ISSUE #7: Replace request by node-fetch
https://github.com/anvilco/node-anvil/blob/master/src/index.js#L1

Cf https://github.com/request/request/issues/3142

### PR DESCRIPTION
**Pull request by newhouse**
## Description of the change

Replace `request` with `node-fetch`. Also:
- add support for streams while we're at it.
- add `User-Agent` option and default.

image

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

Fixes #7 

### Development

- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development
- [ ] No previous tests unrelated to the changed code fail in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached.
- [ ] At least one reviewer has been requested
- [ ] Changes have been reviewed by at least one other engineer
- [ ] The relevant project board has been selected in Projects to auto-link to this pull request

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -29,34 +29,42 @@ class Anvil {
...
-        if (isObject && data.errors) return { statusCode, ...data }
-        else if (isObject && data.message) return { statusCode, errors: [data] }
+        const json = await response.json()
+        const errors = json.errors || (json.message && [json.message])
+
+        return errors ? { statusCode, errors } : { statusCode, ...json }
       }
+
+      const { dataType } = clientOptions
+      const data = dataType === 'stream' ? response.body : await response.buffer()
```
Comment (newhouse) [inline src/index.js:66]: `response.body` is technically a [PassThrough](https://nodejs.org/api/stream.html#stream_class_stream_passthrough), but for all intents and purposes, it's a stream. Not sure how explicit we might want to be about this in our documentation.

File: yarn.lock [hunk 2]
```diff
@@ -155,16 +155,6 @@ ajv@^6.10.0, ajv@^6.10.2:
     json-schema-traverse "^0.4.1"
```
Comment (newhouse) [inline yarn.lock:155]: [All reds!](https://www.youtube.com/watch?v=wouHpNCbqQY)

File: src/index.js [hunk 3]
```diff
@@ -29,34 +29,42 @@ class Anvil {
     this.limiter = new RateLimiter(this.requestLimit, this.requestLimitMS, true)
   }
 
-  fillPDF (pdfTemplateID, payload) {
-    return this.requestREST(`/api/v1/fill/${pdfTemplateID}.pdf`, {
-      method: 'POST',
-      json: payload,
-      encoding: null,
```
Comment (benogle) [inline src/index.js:36]: All good without specifying an encoding? It doesnt try to translate into UTF8 by default?

Comment (newhouse) [reply src/index.js:36]: It is my understanding that for the `request` library, the `encoding: null` option said "I want the response to be a Buffer", which makes sense b/c we wrote the client to deal with buffers as the happy-path response.

With `node-fetch`, you explicitly tell it how you'd like to consume the response via the `response.buffer()` (or similar) calls, and there is no UTF-stuff that happens when dealing with the `stream` or `buffer` types. There is, for example, when you call `response.json()` to consume it.

Anywho, I think we're good and this worked for me no problem locally.

From the `request` documentation:
> `encoding` - encoding to be used on setEncoding of response data. If null, the body is returned as a Buffer. Anything else (including the default value of undefined) will be passed as the encoding parameter to toString() (meaning this is effectively utf8 by default). (Note: if you expect binary data, you should set encoding: null.)

Relevant `node-fetch` docs: https://www.npmjs.com/package/node-fetch#interface-body

**Comment by newhouse** [comment]
@benogle  plz have another look now that I've updated docs and tests...ended up changing a few things.

File: src/index.js [hunk 4]
```diff
@@ -1,8 +1,15 @@
-const request = require('request')
+const fetch = require('node-fetch')
```
Comment (ralbayaty) [inline src/index.js:1]: 🥳

File: src/index.js [hunk 5]
```diff
@@ -1,8 +1,15 @@
...
 
+const { version, description } = require('../package.json')
+
+const DATA_TYPE_STREAM = 'stream'
+const DATA_TYPE_BUFFER = 'buffer'
+const DATA_TYPE_JSON = 'json'
+
 const defaultOptions = {
   baseURL: 'https://app.useanvil.com',
+  userAgent: `${description}/${version}`,
```
Comment (ralbayaty) [inline src/index.js:12]: noice

File: src/index.js [hunk 6]
```diff
@@ -12,12 +19,14 @@ class Anvil {
...
   //   accessToken: <yourAPIKey>, // OR oauth access token
   //   baseURL: 'https://app.useanvil.com'
+  //   userAgent: 'Anvil API Client/2.0.0'
   // }
   constructor (options) {
     if (!options) throw new Error('options are required')
-    this.options = Object.assign({}, defaultOptions, options)
     if (!options.apiKey && !options.accessToken) throw new Error('apiKey or accessToken required')
 
+    this.options = Object.assign({}, defaultOptions, options)
```
Comment (ralbayaty) [inline src/index.js:28]: for a custom supplied `userAgent` would we still want to tack on the `/<version>`? Would let us know which client version they were using. I suppose this isn't used currently though.

File: README.md [hunk 7]
```diff
@@ -42,23 +42,35 @@ const { statusCode, data } = await anvilClient.fillPDF(pdfTemplateID, exampleDat
...
+### Options
+
+Options for the Anvil Client. Defaults are shown after each option key.
+
+```js
+{
+  apiKey: <your_api_key>, // Required. Your API key from your Anvil organization settings
+  baseUrl: 'https://app.useanvil.com', // Optional. String setting the base URL for all client requests
+  userAgent: 'Anvil API Client/<version_number>' // Optional. String setting the User-Agent header in all requests
+}
```
Comment (benogle) [inline README.md:70]: Not sure we want to put these front and center. They probably shouldnt be used. I'd be ok if they were undocumented

Comment (newhouse) [reply README.md]: Agreed.

File: src/index.js (hunk 6 above)
Comment (newhouse) [reply src/index.js:28]: We _could_ do that...but it sort of felt to me like if the User was overriding it that we shouldn't tinker with things. We decided to leave this undocumented for now anyways.

Another thing we could consider is always adding a custom header indicating that it came from our node module.

For now, it's nice to get our client in the user agent header though!

**Pull request merged**